src/pybind/mgr/dashboard/controllers/auth.py

```diff
@@ -28,33 +29,48 @@ class Auth(RESTController):
     """
     Provide authenticates and returns JWT token.
     """
-
     def create(self, username, password):
         user_data = AuthManager.authenticate(username, password)
         user_perms, pwd_expiration_date, pwd_update_required = None, None, None
-        if user_data:
-            user_perms = user_data.get('permissions')
-            pwd_expiration_date = user_data.get('pwdExpirationDate', None)
-            pwd_update_required = user_data.get('pwdUpdateRequired', False)
-
-        if user_perms is not None:
-            logger.debug('Login successful')
-            token = JwtManager.gen_token(username)
-            token = token.decode('utf-8')
-            cherrypy.response.headers['Authorization'] = "Bearer: {}".format(token)
-            return {
-                'token': token,
-                'username': username,
-                'permissions': user_perms,
-                'pwdExpirationDate': pwd_expiration_date,
-                'sso': mgr.SSO_DB.protocol == 'saml2',
-                'pwdUpdateRequired': pwd_update_required
-            }
+        max_attempt = Settings.ACCOUNT_LOCKOUT_ATTEMPTS
+        user_attempts = mgr.ACCESS_CTRL_DB.get_attempt(username)
+        if max_attempt == 0 or user_attempts < max_attempt:
+            if user_data:
+                user_perms = user_data.get('permissions')
+                pwd_expiration_date = user_data.get('pwdExpirationDate', None)
+                pwd_update_required = user_data.get('pwdUpdateRequired', False)
 
-        logger.debug('Login failed')
-        raise DashboardException(msg='Invalid credentials',
-                                 code='invalid_credentials',
-                                 component='auth')
+            if user_perms is not None:
+                logger.info('Login successful: %s', username)
+                mgr.ACCESS_CTRL_DB.save_attempt(username, 0)
+                mgr.ACCESS_CTRL_DB.save()
```

Comment:
I see a potential test-and-set issue here:
* Imagine an attacker launches thousands of login requests at the same time against the same username.
* Cherrypy has only 10 threads for attending requests, so only 10 of these requests will happen in parallel (R_0 - R_9) and the rest of them will remain waiting
* R_0 gets the number of attempts for that username (`user_attempts=0`)
* R_1 gets the number of attempts for that username (`user_attempts=0`)
...
* R_9 gets the number of attemps for that username (`user_attempts=0`)
...
* R_0 to R_9 pass the `if max_attempt == 0 or user_attempts < max_attempt:`, since `user_attempts=0` and they all increment the `user_attempts=1`.
* R_0 stores `user_attempts=1`
* R_1 stores `user_attempts=1`
...
* R_9 stores `user_attempts=1`
* The next 10 requests are accepted by Cherrypy

This means that, under some heavy load, the real number of login attempts allowed could be higher than 10: up to 10*10 (`thread_pool_size * max_attempts`), or even more under certain unusual race conditions (some 'lazy' requests overwrites `user_attempts=1` when a newer request have already stored `user_attempts=2`, delaying even more the blocking).

If you test this with [Apache benchmark](https://httpd.apache.org/docs/2.4/programs/ab.html) tool, I'm almost sure that you'll see probably more than 10 failed logins in the logs prior to see the 'user blocked' log trace.

Reply:
Thank you @epuertat. As per the discussion I've acquired a lock at the time of attempt fetching and released it when the attempt is increased. I will also paste the benchmarking here for you to check it out.

**WITH LOCK IMPLEMENTATION**
```
[nia@localhost ceph]$ ab -n 10000 -c 1000 -A admin:admins http://localhost:4200/#/login
This is ApacheBench, Version 2.3 <$Revision: 1879490 $>
Copyright 1996 Adam Twiss, Zeus Technology Ltd, http://www.zeustech.net/
Licensed to The Apache Software Foundation, http://www.apache.org/

Benchmarking localhost (be patient)
Completed 1000 requests
Completed 2000 requests
Completed 3000 requests
Completed 4000 requests
Completed 5000 requests
Completed 6000 requests
Completed 7000 requests
Completed 8000 requests
Completed 9000 requests
Completed 10000 requests
Finished 10000 requests


Server Software:        
Server Hostname:        localhost
Server Port:            4200

Document Path:          /#/login
Document Length:        1059 bytes

Concurrency Level:      1000
Time taken for tests:   3.888 seconds
Complete requests:      10000
Failed requests:        0
Total transferred:      13160000 bytes
HTML transferred:       10590000 bytes
Requests per second:    2572.02 [#/sec] (mean)
Time per request:       388.800 [ms] (mean)
Time per request:       0.389 [ms] (mean, across all concurrent requests)
Transfer rate:          3305.45 [Kbytes/sec] received

Connection Times (ms)
              min  mean[+/-sd] median   max
Connect:        0   17   8.4     17      36
Processing:    36  324 283.1    253    1396
Waiting:        1  211 284.7    118    1332
Total:         36  341 281.9    274    1411

Percentage of the requests served within a certain time (ms)
  50%    274
  66%    308
  75%    329
  80%    346
  90%    477
  95%   1304
  98%   1394
  99%   1403
 100%   1411 (longest request)
```

**WITHOUT LOCK IMPLEMENTATION**
```
[nia@localhost ceph-dev]$ ab -n 10000 -c 1000 -A admin:admins http://localhost:4200/#/login
This is ApacheBench, Version 2.3 <$Revision: 1879490 $>
Copyright 1996 Adam Twiss, Zeus Technology Ltd, http://www.zeustech.net/
Licensed to The Apache Software Foundation, http://www.apache.org/

Benchmarking localhost (be patient)
Completed 1000 requests
Completed 2000 requests
Completed 3000 requests
Completed 4000 requests
Completed 5000 requests
Completed 6000 requests
Completed 7000 requests
Completed 8000 requests
Completed 9000 requests
Completed 10000 requests
Finished 10000 requests


Server Software:        
Server Hostname:        localhost
Server Port:            4200

Document Path:          /#/login
Document Length:        1059 bytes

Concurrency Level:      1000
Time taken for tests:   2.557 seconds
Complete requests:      10000
Failed requests:        0
Total transferred:      13160000 bytes
HTML transferred:       10590000 bytes
Requests per second:    3910.66 [#/sec] (mean)
Time per request:       255.712 [ms] (mean)
Time per request:       0.256 [ms] (mean, across all concurrent requests)
Transfer rate:          5025.81 [Kbytes/sec] received

Connection Times (ms)
              min  mean[+/-sd] median   max
Connect:        0   12   9.0      9      57
Processing:    13  233 195.2    205    1271
Waiting:        1  176 197.6    142    1239
Total:         58  245 194.7    222    1290

Percentage of the requests served within a certain time (ms)
  50%    222
  66%    250
  75%    252
  80%    254
  90%    258
  95%    263
  98%   1268
  99%   1279
 100%   1290 (longest request)
```